### ISSUE TITLE
Horust hangs forever when received SIGTERM before all services go up and running.

### ISSUE THREAD

**Issue by frondeus** [FedericoPonzi/Horust#58]
#### Version Information:

* Version: v0.1.3

First noticed the bug on my modified branch (see PR-draft #57), however, the same bug occurs on fresh `master` and when my friend installed `horust` via `cargo install`.

* OS: Linux
* CPU: x64

#### Description:
Scenario:

* I start horust services
* Not all services are running, at least one is still "Starting",
* I have to quickly kill them all (because for example I made a mistake in configuration and I don't want to wait till service goes up)

Then Horust will terminate all already running services, but waits forever for this late service which is never changing its state from `Starting` to `Started`.

It means that in the main loop Horust checks whether all services are `Finished` or `Failed`, but this one service will just wait forever in `Starting` state. I guess for some reason it never receives `SpawnFailed` event?

The quickest way to test it is to add `start-delay = "2s"` in `service.toml`.

I made some `eprintln!` and tried to debug it:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
[2021-04-29T07:51:21Z INFO  horust] Loading services from directories:
    * ./core
    * ./extra
[2021-04-29T07:51:21Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:21Z INFO  horust::horust::supervisor] Applying events... [
    Run("1.toml"), Run("2.toml"), Run("3.toml"), Run("4.toml"), Run("5.toml"), Run("6.toml"), Run("7.toml"), Run("8.toml"), Run("9.toml"), Run("10.toml")
]
[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
^C[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:22Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("6.toml")
[2021-04-29T07:51:22Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:22Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("7.toml")
[2021-04-29T07:51:23Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:23Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("8.toml")
[2021-04-29T07:51:23Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:23Z WARN  horust::horust::supervisor] 1. SIGTERM received
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: SpawnFailed("4.toml")
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("2.toml", Pid(66623))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("5.toml", Pid(66624))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("1.toml", Pid(66625))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("10.toml", Pid(66626))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("3.toml", Pid(66627))
SPAWN_FOR_EXEC_HANDLER_LOOP BROKEN: PidChanged("9.toml", Pid(66628))
[2021-04-29T07:51:23Z INFO  horust::horust::supervisor] Applying events... [
    PidChanged("2.toml", Pid(66623)),
    PidChanged("5.toml", Pid(66624)),
    PidChanged("1.toml", Pid(66625)),
    PidChanged("10.toml", Pid(66626)),
    PidChanged("3.toml", Pid(66627)),
    PidChanged("9.toml", Pid(66628))
]
[2021-04-29T07:51:23Z WARN  horust::horust::supervisor] 1. SIGTERM received
PID CHANGED: "2.toml" - 66623
PID CHANGED: "5.toml" - 66624
PID CHANGED: "1.toml" - 66625
PID CHANGED: "10.toml" - 66626
PID CHANGED: "3.toml" - 66627
PID CHANGED: "9.toml" - 66628
[2021-04-29T07:51:24Z INFO  horust::horust::supervisor] Applying events... [
    ShuttingDownInitiated(Gracefuly),
    StatusChanged("2.toml", Started),
    StatusChanged("5.toml", Started),
    StatusChanged("1.toml", Started),
    StatusChanged("10.toml", Started),
    StatusChanged("3.toml", Started),
    StatusChanged("9.toml", Started)
]
[2021-04-29T07:51:24Z WARN  horust::horust::supervisor] 1. SIGTERM received
[2021-04-29T07:51:24Z WARN  horust::horust::supervisor] Gracefully stopping...
[2021-04-29T07:51:24Z INFO  horust::horust::supervisor] Applying events... [
    ShuttingDownInitiated(Gracefuly),
    StatusUpdate("1.toml", InKilling),
    Kill("1.toml"),
    StatusUpdate("2.toml", InKilling),
    Kill("2.toml"),
    StatusUpdate("3.toml", InKilling),
    Kill("3.toml"),
    StatusUpdate("5.toml", InKilling),
    Kill("5.toml"),
    StatusUpdate("11.toml", Finished),
    StatusUpdate("9.toml", InKilling),
    Kill("9.toml"),
    StatusUpdate("10.toml", InKilling),
    Kill("10.toml")
]
[2021-04-29T07:51:24Z WARN  horust::horust::supervisor] Gracefully stopping...
[2021-04-29T07:51:24Z INFO  horust::horust::supervisor] Applying events... [
    StatusChanged("1.toml", InKilling),
    StatusChanged("2.toml", InKilling),
    StatusChanged("3.toml", InKilling),
    StatusChanged("5.toml", InKilling),
    StatusChanged("11.toml", Finished),
    StatusChanged("9.toml", InKilling),
    StatusChanged("10.toml", InKilling)
]
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:25Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:26Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:26Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:26Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:27Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:27Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:27Z INFO  horust::horust::supervisor] Applying events... [
    ForceKill("1.toml"),
    ForceKill("2.toml"),
    ForceKill("3.toml"),
    ForceKill("5.toml"),
    ForceKill("9.toml"),
    ForceKill("10.toml")
]
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... [
    StatusChanged("1.toml", Failed),
    StatusChanged("2.toml", Failed),
    StatusChanged("3.toml", Failed),
    StatusChanged("5.toml", Failed),
    StatusChanged("9.toml", Failed),
    StatusChanged("10.toml", Failed),
    StatusUpdate("1.toml", FinishedFailed),
    StatusUpdate("2.toml", FinishedFailed),
    StatusUpdate("3.toml", FinishedFailed),
    StatusUpdate("5.toml", FinishedFailed),
    StatusUpdate("9.toml", FinishedFailed),
    StatusUpdate("10.toml", FinishedFailed)
]
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... [
    StatusChanged("1.toml", FinishedFailed),
    StatusChanged("2.toml", FinishedFailed),
    StatusChanged("3.toml", FinishedFailed),
    StatusChanged("5.toml", FinishedFailed),
    StatusChanged("9.toml", FinishedFailed),
    StatusChanged("10.toml", FinishedFailed)
]
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... []
[2021-04-29T07:51:28Z INFO  horust::horust::supervisor] Applying events... []
At this point there are no new events forever (loop with empty Applying events... []
```
My modifications:
* Added Graceful/Forceful stop - but as I wrote, it isn't the cause as the bug occurs also on fresh `main` branch.
* Added `SPAWN_FOR_EXEC_HANDLER_LOOP` eprintln! in `process_spawner` just after `loop` ends (before `bus.send_event(ev)`. Thanks to that, we can see that `SpawnFailed` event is created!
* I added two extra `exprintln!`: in `supervisor/mod.rs::handle_event()` to see if we ever process `SpawnFailed` or `PidChanged`. One is processed, another isnt.
* So, I changed log level for `Applying events` to `info` -> We can see that `SpawnFailed` event is never received!
* But `PidChanged` is received. Why?

Extra conclusion:
Whats interesting, one service ("11.toml") had `start-delay = 10s` instead of `2s`, and it had `start-after=["8.toml"]`.
So horust correctly were waiting till `8.toml` goes up, but its hanged on `Starting` (Spawn Failed has not been processed), therefore `11.toml` is in `Initial` state, so `service_handler.rs` changed its status to `Finished` correctly.

The only way to kill horust then is `killall horust -9`

**Comment by frondeus** [comment]
I also tried to debug `Bus` - I added `Debug` trait to `where` clause, and wrote `dbg!(&ev.clone().into_payload()` for each received message in the `dispatch()` function... And it received `SpawnFailed`!
I didn't fully read the code yet but my assumption is:
* Either it didn't send the event to supervisor.. because supervisor was not running yet
* Or there is some strange bug with channels.

**Comment by frondeus** [comment]
Supervisor is running, and dispatch is sending data to the channel. Whats more, `sender.send()` for supervisor channel returns `Ok()`... But supervisor never receives this data.

Honestly, I'm surprised and I don't know why it doesn't work.
I guess it might be related to `std::thread::sleep` in supervisor, because when I commented this line, it started working. This might be a red herring tho, because I didn't see any `SpawnFailed` sent at all.

**Comment by frondeus** [comment]
Okay, I think I found the root cause.

In `supervisor/mod.rs::handle_event()` there is:
```rust
process_spawner::spawn_fork_exec_handler(
    service_handler.service().clone(),
    backoff,
    self.repo.bus.clone()
    ^^^^^^^^^^^^^^^^^^^ - this is wrong.
);
```

By cloning BusConnector, horust creates a round-robin dispatching of these messages.
ProcessSpawner shares the same ID in the bus, plus it shares the same channel. It means sometimes process spawner (or even another thread, cause I see `.clone()` on BusConnector in many places) takes the message and the message is never processed by the supervisor.

Instead, the bus itself needs to be somehow shared (for example by using `Arc<Mutex>` inside of every `BusConnector`), so instead of cloning there would be `bus_connector.join_bus()`).

**Comment by FedericoPonzi** [comment]
Thanks for reporting this! 

I've tried to add `start-delay = "2s"` in `test/utils/mod.rs:33` and run `test_stress_test_chained_services` test, but I wasn't able to reproduce the issue. A part from this, I think your last message is really on point and that should really be a `join_bus` and not a clone.

I'm wondering because you didn't mention that, if you replace the clone with a `join_bus` call, is the issue resolved?

**Comment by frondeus** [comment]
Sorry, I think I wrote not enough info how to test it:
I tested it only manually.
I added `start-delay` to most of my services and then immediately after horust started I pressed `CTRL+C` (once, or several times, it does not matter).

 I saw in logs that after Horust received `Event::ShuttingDownInitiated` it didn't handle `SpawnFailed` correctly.


Anyways - today I spent all my work time for OSS projects therefore I'll check it tomorrow. But it isn't simple `join_bus()`. There is cloned and propagated `BusConnector` struct, while `join_bus()` is a method in `Bus`. The `Bus` instance lives in `horust/mod.rs::Horust::run()` method and its not propagated deeper into the systems.

Therefore to change this id need to make a larger refactor of `Bus` and `BusConnector`. I'll try to touch it tomorrow.
